### PR TITLE
fanout: init at unstable-2023-07-21

### PR DESCRIPTION
**Pull request by therishidesai**
###### Description of changes

Add the [fanout](https://github.com/bob-linuxtoys/fanout/tree/master) kernel module into linuxPackages.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/fanout/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  patches = [
+    ./use_modules_install.patch
+    ./remove_udev.patch
+  ];
+
+  hardeningDisable = [ "format" "pic" ];
+
+  nativeBuildInputs = [ kmod ] ++ kernel.moduleBuildDependencies;
+
+  outputs = [ "out" ];
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/fanout/default.nix:27]: Isn't this the default?

File: pkgs/os-specific/linux/fanout/default.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    "KERNELDIR=${kernel.dev}/lib/modules/${kernel.modDirVersion}/build"
+  ];
+
+  meta = with lib; {
+    description = "Kernel-based publish-subscribe system";
+    homepage = "https://github.com/bob-linuxtoys/fanout";
+    license = licenses.gpl2Only;
+    maintainers = with maintainers; [ therishidesai ];
+    platforms = platforms.linux;
+    outputsToInstall = [ "out" ];
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/fanout/default.nix:39]: Same here

File: pkgs/os-specific/linux/fanout/remove_udev.patch [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+--- a/Makefile
++++ b/Makefile
+@@ -16,6 +16,6 @@ install:
+ 	depmod -a
+ endif
+ 
+-udev:
+-	cp udev.d/fanout.rules /etc/udev/rules.d/90-fanout.rules
+-	udevadm control --reload-rules
++# udev:
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/fanout/remove_udev.patch:12]: Is the target is now missing, doesn't this mean it was unuses anyway?

File: pkgs/os-specific/linux/fanout/use_modules_install.patch [hunk 4]
```diff
@@ -0,0 +1,24 @@
+diff --git a/Makefile b/Makefile
+index 6fd8e28..9da6ebe 100644
+--- a/Makefile
++++ b/Makefile
+@@ -3,7 +3,7 @@ ifneq ($(KERNELRELEASE),)
+ else
+ 	KERNELDIR ?= /lib/modules/$(shell uname -r)/build
+ 	PWD := $(shell pwd)
+-
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/fanout/use_modules_install.patch:9]: We can delete this hunk, right?

**Comment by therishidesai** [comment]
@SuperSandro2000 Thanks for the feedback. Cleaned up the PR

**Comment by therishidesai** [comment]
Got a makefile fix into the upstream fanout repo so was able to remove the patch file

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 244599` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.hardkernel_4_14.fanout</li>
    <li>linuxPackages_hardkernel_latest.fanout</li>
  </ul>
</details>
<details>
  <summary>15 packages failed to build:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_14.fanout</li>
    <li>linuxPackages_4_14_hardened.fanout (linuxKernel.packages.linux_4_14_hardened.fanout)</li>
    <li>linuxKernel.packages.linux_4_19.fanout</li>
    <li>linuxPackages_4_19_hardened.fanout (linuxKernel.packages.linux_4_19_hardened.fanout)</li>
    <li>linuxKernel.packages.linux_5_4.fanout</li>
    <li>linuxPackages_5_4_hardened.fanout (linuxKernel.packages.linux_5_4_hardened.fanout)</li>
    <li>linuxKernel.packages.linux_6_3.fanout</li>
    <li>linuxPackages_6_3_hardened.fanout (linuxKernel.packages.linux_6_3_hardened.fanout)</li>
    <li>linuxPackages_latest.fanout (linuxKernel.packages.linux_6_4.fanout)</li>
    <li>linuxPackages_6_4_hardened.fanout (linuxKernel.packages.linux_6_4_hardened.fanout)</li>
    <li>linuxPackages_latest-libre.fanout (linuxKernel.packages.linux_latest_libre.fanout)</li>
    <li>linuxPackages_lqx.fanout (linuxKernel.packages.linux_lqx.fanout)</li>
    <li>linuxPackages_testing_bcachefs.fanout (linuxKernel.packages.linux_testing_bcachefs.fanout)</li>
    <li>linuxPackages_xanmod_latest.fanout (linuxKernel.packages.linux_xanmod_latest.fanout ,linuxPackages_xanmod_stable.fanout)</li>
    <li>linuxPackages_zen.fanout (linuxKernel.packages.linux_zen.fanout)</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_5_10.fanout</li>
    <li>linuxPackages_5_10_hardened.fanout (linuxKernel.packages.linux_5_10_hardened.fanout)</li>
    <li>linuxKernel.packages.linux_5_15.fanout</li>
    <li>linuxPackages_5_15_hardened.fanout (linuxKernel.packages.linux_5_15_hardened.fanout)</li>
    <li>linuxPackages.fanout (linuxKernel.packages.linux_6_1.fanout)</li>
    <li>linuxPackages_hardened.fanout (linuxPackages_6_1_hardened.fanout)</li>
    <li>linuxPackages-libre.fanout (linuxKernel.packages.linux_libre.fanout)</li>
    <li>linuxPackages_xanmod.fanout (linuxKernel.packages.linux_xanmod.fanout)</li>
  </ul>
</details>

**Comment by therishidesai** [comment]
```
make -C /nix/store/9lph03skr2gb4412shxla4ij9m6wspmc-linux-4.19.288-dev/lib/modules/4.19.288/build M=/build/source modules
  CC [M]  /build/source/fanout.o
/build/source/fanout.c:28:12: fatal error: linux/device/class.h: No such file or directory
   28 | #  include <linux/device/class.h>
      |            ^~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[4]: *** [/nix/store/9lph03skr2gb4412shxla4ij9m6wspmc-linux-4.19.288-dev/lib/modules/4.19.288/source/scripts/Makefile.build:310: /build/source/fanout.o] Error 1
make[3]: *** [/nix/store/9lph03skr2gb4412shxla4ij9m6wspmc-linux-4.19.288-dev/lib/modules/4.19.288/source/Makefile:1562: _module_/build/source] Error 2
make[2]: *** [Makefile:153: sub-make] Error 2
make[1]: *** [Makefile:24: __sub-make] Error 2
make: *** [Makefile:8: all] Error 2
```

It seems to be failing with this on some of the kernels

**Comment by therishidesai** [comment]
https://github.com/bob-linuxtoys/fanout/blob/master/fanout.c#L16

Looks like the auto creation of the dev device is where that include is needed. Not a necessary feature of the module so just gonna add a patch file that comments out the line above so it can build on more kernels

**Comment by therishidesai** [comment]
Seems to be working now:

Result of `nixpkgs-review pr 244599` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>23 packages built:</summary>
  <ul>
    <li>linuxKernel.packages.linux_4_14.fanout</li>
    <li>linuxKernel.packages.linux_4_14_hardened.fanout</li>
    <li>linuxKernel.packages.linux_4_19.fanout</li>
    <li>linuxKernel.packages.linux_4_19_hardened.fanout</li>
    <li>linuxKernel.packages.linux_5_10.fanout</li>
    <li>linuxKernel.packages.linux_5_10_hardened.fanout</li>
    <li>linuxKernel.packages.linux_5_15.fanout</li>
    <li>linuxKernel.packages.linux_5_15_hardened.fanout</li>
    <li>linuxKernel.packages.linux_5_4.fanout</li>
    <li>linuxKernel.packages.linux_5_4_hardened.fanout</li>
    <li>linuxKernel.packages.linux_6_1.fanout</li>
    <li>linuxKernel.packages.linux_hardened.fanout (linuxKernel.packages.linux_6_1_hardened.fanout)</li>
    <li>linuxKernel.packages.linux_6_3.fanout</li>
    <li>linuxKernel.packages.linux_6_3_hardened.fanout</li>
    <li>linuxKernel.packages.linux_6_4.fanout</li>
    <li>linuxKernel.packages.linux_6_4_hardened.fanout</li>
    <li>linuxKernel.packages.linux_latest_libre.fanout</li>
    <li>linuxKernel.packages.linux_libre.fanout</li>
    <li>linuxKernel.packages.linux_lqx.fanout</li>
    <li>linuxKernel.packages.linux_testing_bcachefs.fanout</li>
    <li>linuxKernel.packages.linux_xanmod.fanout</li>
    <li>linuxKernel.packages.linux_xanmod_latest.fanout (linuxKernel.packages.linux_xanmod_stable.fanout)</li>
    <li>linuxKernel.packages.linux_zen.fanout</li>
  </ul>
</details>

**Pull request merged**